### PR TITLE
Put the progressbar back in the Npm UI

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Add the progress bar back when installing npm packages from the npm UI.

This fixes:

https://developercommunity.visualstudio.com/content/problem/154181/no-visual-feedback-when-installing-npm-package.html

### CODE REVIEW

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 1]
```diff
@@ -218,16 +233,11 @@ private void StartFilter()
 
             if (filtered.Any())
             {
-                IRootPackage rootPackage = null;
-                var controller = this.npmController;
-                if (controller != null)
-                {
-                    rootPackage = controller.RootPackage;
-                }
+                var rootPackage = this.npmController?.RootPackage;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: >rootPackage [](start = 20, length = 11)

Now that it's only consumed once, you could probably just inline it.

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 2]
```diff
@@ -357,17 +367,64 @@ public PackageCatalogEntryViewModel SelectedPackage
...
+            get
+            {
+                return this.isExecutingCommand;
+            }
+            set
+            {
+                this.isExecutingCommand = value;
+
+                OnPropertyChanged();
+                OnPropertyChanged("ExecutionProgressVisibility");
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: >ExecutionProgressVisibility [](start = 35, length = 27)

`nameof`

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 3]
```diff
@@ -357,17 +367,64 @@ public PackageCatalogEntryViewModel SelectedPackage
...
-                handler(this, new PropertyChangedEventArgs(propertyName));
+                if (disposing)
+                {
+                    this.filterTimer?.Dispose();
+                }
+
+                var controller = this.npmController;
+                if (controller != null)
+                {
+                    this.npmController.FinishedRefresh -= this.NpmController_FinishedRefresh;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: >this.npmController [](start = 20, length = 18)

`controller`.  Was it extracted for fear of a race?

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 4]
```diff
@@ -357,17 +367,64 @@ public PackageCatalogEntryViewModel SelectedPackage
...
-            if (handler != null)
+            this.PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposed)
             {
-                handler(this, new PropertyChangedEventArgs(propertyName));
+                if (disposing)
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: >if (disposing) [](start = 16, length = 14)

Why doesn't this apply to the event changes as well?

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs (hunk 3 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: no the compiler complained

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 5]
```diff
@@ -357,17 +367,64 @@ public PackageCatalogEntryViewModel SelectedPackage
...
+                }
+                disposed = true;
             }
         }
+
+        // This code added to correctly implement the disposable pattern.
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+            Dispose(true);
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: `GC.SuppressFinalize(this);`?

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 6]
```diff
@@ -357,17 +367,64 @@ public PackageCatalogEntryViewModel SelectedPackage
...
+
+                var controller = this.npmController;
+                if (controller != null)
+                {
+                    this.npmController.FinishedRefresh -= this.NpmController_FinishedRefresh;
+                }
+                var worker = this.npmWorker;
+                if (worker != null)
+                {
+                    this.npmWorker.CommandStarted -= this.NpmWorker_CommandStarted;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: >this.npmWorker [](start = 20, length = 14)

`worker`

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 7]
```diff
@@ -357,17 +367,64 @@ public PackageCatalogEntryViewModel SelectedPackage
...
+        {
+            if (!disposed)
             {
-                handler(this, new PropertyChangedEventArgs(propertyName));
+                if (disposing)
+                {
+                    this.filterTimer?.Dispose();
+                }
+
+                var controller = this.npmController;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: >var controller = this.npmController [](start = 16, length = 35)

The reason not to just set the corresponding property to null is that we don't want to fire a property changed event?

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs (hunk 2 above)
Comment (amcasey) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: Strictly speaking, this should only happen if the value actually changed.

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs (hunk 3 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: mostly worried about the VS shutdown/solution close scenario

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs (hunk 5 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: only required when you override the finalizer

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs (hunk 7 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: yep

**Pull request merged**